src/realm/obj.cpp

```diff
@@ -829,34 +839,136 @@ void out_mixed(std::ostream& out, const Mixed& val)
+            out << timeMillis;
+            out << "\"}}";
+            break;
+        }
+        case type_Decimal:
+            out << "{\"$numberDecimal\": \"";
+            out << val.get<Decimal128>();
+            out << "\"}";
+            break;
+        case type_ObjectId:
+            out << "{\"$oid\": \"";
+            out << val.get<ObjectId>();
+            out << "\"}";
+            break;
+        case type_TypedLink: {
+            auto link = val.get<ObjLink>();
+            out << "{ \"$link\": { \"table\": \"" << link.get_table_key() << "\", \"key\": ";
+            out_mixed_xjson(out, link.get_obj_key());
+            out << "}}";
+            break;
+        }
+        case type_Link:
+        case type_LinkList:
+        case type_OldDateTime:
+        case type_Mixed:
+        case type_OldTable:
+            break;
+    }
+}
 
+void out_mixed(std::ostream& out, const Mixed& val, JSONOutputMode output_mode)
+{
+    switch (output_mode) {
+        case output_mode_xjson: {
+            out_mixed_xjson(out, val);
+            return;
+        }
+        case output_mode_json: {
+            out_mixed_json(out, val);
+        }
+    }
+}
+
+} // anonymous namespace
 void Obj::to_json(std::ostream& out, size_t link_depth, std::map<std::string, std::string>& renames,
-                  std::vector<ColKey>& followed) const
+                  std::vector<ColKey>& followed, JSONOutputMode output_mode) const
 {
     StringData name = "_key";
+    bool prefixComma = false;
     if (renames[name] != "")
         name = renames[name];
     out << "{";
-    out << "\"" << name << "\":" << this->m_key.value;
+    if (output_mode != output_mode_xjson) {
+        prefixComma = true;
+        out << "\"" << name << "\":" << this->m_key.value;
+    }
+
     auto col_keys = m_table->get_column_keys();
     for (auto ck : col_keys) {
         name = m_table->get_column_name(ck);
         auto type = ck.get_type();
         if (renames[name] != "")
             name = renames[name];
 
-        out << ",\"" << name << "\":";
+        if (prefixComma)
+            out << ",";
+        out << "\"" << name << "\":";
+        prefixComma = true;
 
         if (ck.get_attrs().test(col_attr_List)) {
             if (type == col_type_LinkList) {
                 TableRef target_table = get_target_table(ck);
+                auto primary_key_coll = target_table->get_primary_key_column();
                 auto ll = get_linklist(ck);
                 auto sz = ll.size();
 
-                if (!target_table->is_embedded() &&
-                    ((link_depth == 0) || (link_depth == not_found &&
-                                           std::find(followed.begin(), followed.end(), ck) != followed.end()))) {
+                if (output_mode == output_mode_xjson && !target_table->is_embedded() && primary_key_coll) {
+                    out << "{ \"$linkList\": { \"table\": \"" << get_target_table(ck)->get_name()
```

Comment:
If this is a list of links, won't this render as {"$linkList: { "table": "foo", keys: [ { $link: { table: "foo", id: "bar" } } ] } }. That seems pretty verbose! Is this just so the fuzzer can verify that it's actually a list rather than an array?

Reply:
This outputs the primary key only, so it will look like {"$linkList: { "table": "foo", keys: [ "bar", "biz", "baz"] } }